File: java/src/main/java/com/github/apstndb/spanvalue/ClientTypeAdapter.java
package com.github.apstndb.spanvalue;

import com.google.spanner.v1.StructType;
import com.google.spanner.v1.Type;
import com.google.spanner.v1.TypeAnnotationCode;
import com.google.spanner.v1.TypeCode;
import java.lang.reflect.Method;
import java.util.List;

/** Adapt high-level client types to wire {@link com.google.spanner.v1.Type}. */
public final class ClientTypeAdapter {
  private ClientTypeAdapter() {}

  /**
   * Adapt a {@code com.google.cloud.spanner.Type} (or wire {@link Type}) to wire {@link Type}.
   *
   * <p>The main library does not depend on {@code google-cloud-spanner}; client types are handled
   * via reflection. Wire {@link Type} objects are returned unchanged.
   */
  public static Type adapt(Object clientType) {
    if (clientType == null) {
      throw new IllegalArgumentException("clientType must not be null");
    }
    if (clientType instanceof Type wireType) {
      return wireType;
    }
    if (!"com.google.cloud.spanner.Type".equals(clientType.getClass().getName())) {
      throw new IllegalArgumentException(
          "unsupported client type: " + clientType.getClass().getName());
    }
    return adaptCloudSpannerType(clientType);
  }

  private static Type adaptCloudSpannerType(Object clientType) {
    try {
      Object codeObj = invoke(clientType, "getCode");
      TypeCode wireCode = (TypeCode) invokePackagePrivate(codeObj, "getTypeCode");
      Type.Builder builder = Type.newBuilder().setCode(wireCode);

      Object annObj = invokePackagePrivate(codeObj, "getTypeAnnotationCode");
      if (annObj instanceof TypeAnnotationCode ann
          && ann != TypeAnnotationCode.TYPE_ANNOTATION_CODE_UNSPECIFIED) {
        builder.setTypeAnnotation(ann);
      }

      if (wireCode == TypeCode.PROTO || wireCode == TypeCode.ENUM) {
        String fqn = (String) invoke(clientType, "getProtoTypeFqn");
        if (fqn != null && !fqn.isEmpty()) {
          builder.setProtoTypeFqn(fqn);
        }
      }

      if (wireCode == TypeCode.ARRAY) {
        Object elem = invoke(clientType, "getArrayElementType");
        if (elem != null) {
          builder.setArrayElementType(adapt(elem));
        }
      }

      if (wireCode == TypeCode.STRUCT) {
        @SuppressWarnings("unchecked")
        List<Object> structFields = (List<Object>) invoke(clientType, "getStructFields");
        if (structFields != null && !structFields.isEmpty()) {
          StructType.Builder structBuilder = StructType.newBuilder();
          for (Object field : structFields) {
            String name = (String) invoke(field, "getName");
            Object fieldType = invoke(field, "getType");
            structBuilder.addFields(
                StructType.Field.newBuilder()
                    .setName(name == null ? "" : name)
                    .setType(adapt(fieldType))
                    .build());
          }
          builder.setStructType(structBuilder.build());
        }
      }

      return builder.build();
    } catch (ReflectiveOperationException e) {
      throw new IllegalArgumentException("failed to adapt client type", e);
    }
  }

  private static Object invoke(Object target, String methodName) throws ReflectiveOperationException {
    Method method = target.getClass().getMethod(methodName);
    return method.invoke(target);
  }

  private static Object invokePackagePrivate(Object target, String methodName)
      throws ReflectiveOperationException {
    Method method = target.getClass().getDeclaredMethod(methodName);
    method.setAccessible(true);
    return method.invoke(target);
  }
}

File: nodejs/lib/client-type-adapter.js
/** Duck-typed adapters for @google-cloud/spanner structural types. */

import {
  TypeAnnotationCode,
  TYPE_ANNOTATION_NAMES,
  typeCodeName,
} from './codes.js';
import {
  arrayElementType,
  fieldName,
  fieldType,
  protoTypeFqn,
  structFields,
  structType,
  typeAnnotation,
  typeCode,
} from './proto.js';

/**
 * @param {unknown} clientType
 * @returns {Record<string, unknown>}
 */
export function adaptClientType(clientType) {
  if (clientType == null) {
    return { code: 'TYPE_CODE_UNSPECIFIED' };
  }

  const code = typeCode(clientType);
  const name = typeCodeName(code);
  /** @type {Record<string, unknown>} */
  const wire = { code: name ?? code };

  const ann = typeAnnotation(clientType);
  if (ann != null && ann !== TypeAnnotationCode.TYPE_ANNOTATION_CODE_UNSPECIFIED) {
    wire.typeAnnotation = TYPE_ANNOTATION_NAMES[ann] ?? ann;
  }

  const fqn = protoTypeFqn(clientType);
  if (fqn) {
    wire.protoTypeFqn = fqn;
  }

  const elem = arrayElementType(clientType);
  if (elem != null) {
    wire.arrayElementType = adaptClientType(elem);
  }

  const st = structType(clientType);
  if (st != null) {
    wire.structType = {
      fields: structFields(clientType).map((field) => {
        /** @type {{ type: Record<string, unknown>, name?: string }} */
        const out = { type: adaptClientType(fieldType(field)) };
        const name = fieldName(field);
        if (name) {
          out.name = name;
        }
        return out;
      }),
    };
  }

  return wire;
}
